stl/inc/iterator

```diff
@@ -511,6 +511,279 @@ private:
+        return _Current[_Diff];
+    }
+
+    // [counted.iter.nav]
+    constexpr counted_iterator& operator++() {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Length > 0, "counted_iterator increment beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        ++_Current;
+        --_Length;
+        return *this;
+    }
+
+    constexpr decltype(auto) operator++(int) {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Length > 0, "counted_iterator increment beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        --_Length;
+        _TRY_BEGIN
+        return _Current++;
+
+        _CATCH_ALL
+        ++_Length;
+
+        _RERAISE;
+        _CATCH_END
+    }
+
+    constexpr counted_iterator operator++(int) requires forward_iterator<_Iter> {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Length > 0, "counted_iterator increment beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        counted_iterator _Tmp = *this;
+        ++_Current;
+        --_Length;
+        return _Tmp;
+    }
+
+    constexpr counted_iterator& operator--() requires bidirectional_iterator<_Iter> {
+        --_Current;
+        ++_Length;
+        return *this;
+    }
+
+    constexpr counted_iterator operator--(int) requires bidirectional_iterator<_Iter> {
+        counted_iterator _Tmp = *this;
+        --_Current;
+        ++_Length;
+        return _Tmp;
+    }
+
+    _NODISCARD constexpr counted_iterator operator+(const iter_difference_t<_Iter> _Diff) const
+        requires random_access_iterator<_Iter> {
+        return counted_iterator{_Current + _Diff, static_cast<iter_difference_t<_Iter>>(_Length - _Diff)};
+    }
+
+    _NODISCARD friend constexpr counted_iterator operator+(
+        const iter_difference_t<_Iter> _Diff, const counted_iterator& _Right) requires random_access_iterator<_Iter> {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Diff <= _Right.count(), "counted_iterator seek beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return _Right + _Diff;
+    }
+
+    constexpr counted_iterator& operator+=(
+        const iter_difference_t<_Iter> _Diff) requires random_access_iterator<_Iter> {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Diff <= _Length, "counted_iterator increment beyond end of range");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        _Current += _Diff;
+        _Length -= _Diff;
+        return *this;
+    }
+
+    _NODISCARD constexpr counted_iterator operator-(const iter_difference_t<_Iter> _Diff) const
+        requires random_access_iterator<_Iter> {
+        return counted_iterator{_Current - _Diff, static_cast<iter_difference_t<_Iter>>(_Length + _Diff)};
+    }
+
+    template <common_with<_Iter> _Other>
+    _NODISCARD friend constexpr iter_difference_t<_Other> operator-(
+        const counted_iterator& _Left, const counted_iterator<_Other>& _Right) noexcept /* strengthened */ {
+        _Verify_range(_Left, _Right);

```

Comment:
(See discussion starting at https://github.com/microsoft/STL/pull/1083#discussion_r460287601.)

Having now read the entire code (which I hadn't when I made this comment) I see you're doing something different than our conventional `_Verify_range`. I think that thing is necessary but not sufficient to implement the required semantics of `_Verify_range`, and should therefore have a different name like `_Same_sequence`. To be precise, I think `_Verify_range(_Left, _Right)` needs to require `_Left.count() >= _Right.count() && _Same_sequence(_Left, _Right)` to ensure that `[_Left, _Right)` is a valid range.

`_Same_sequence`, however, is pretty expensive: I don't think we need to advance both iterators to the end of their travel to ensure they are over the same sequence. If `next(_Left._Current, _Left._Length) == next(_Right._Current, _Right._Length())`, then `next(_Left._Current, _Left._Length - 1) == next(_Right._Current, _Right._Length() - 1)` (assuming the underlying sequence is at least 1 element long). Of course this generalizes for all positive lengths so we simply need to show that `next(_Left._Current, x) == next(_Right._Current, y)` where `x` is `max(_Left._Length - _Right._Length, 0)` and `y` is `max(_Right._Length - _Left._Length, 0)`. IOW:
```c++
if (_Left._Length >= _Right._Length) {
    assert(next(_Left._Current, _Left._Length - _Right._Length) == _Right._Current);
} else {
    assert(_Left._Current == next(_Right._Current, _Right._Length - _Left._Length));
}
```
verifies "iterators over the same sequence." For `_Verify_range`, where we know `_Left` must precede `_Right`, we can simply `assert((_Left._Length >= _Right._Length && next(_Left._Current, _Left._Length - _Right._Length) == _Right._Current)`.

> My problem is what I should do with iterators of different type?

All of the operations that accept two iterators of different type require their types `T` and `U` to model `common_with<T, U>`, so you can explicitly convert (`static_cast`) them to `common_type<T, U>` before comparing for equality.

Reply:
Thanks a lot for the help, you are awesome!

Gave it a shot

Reply:
Adopted to it

Reply:
And that crashes the compiler ;(